python/ray/serve/_private/application_state.py

```diff
@@ -332,32 +316,79 @@ def list_deployment_details(self, name: str) -> Dict[str, DeploymentDetails]:
-        This is used for holding the deploy_obj_ref which is created by run_graph method
         """
-        if (
-            name in self._application_states
-            and self._application_states[name].deploy_obj_ref
-        ):
+        Create application state.
+
+        This is used for holding the build_app_obj_ref, which is the object ref for the
+        build_application task.
+        """
+        if name in self._application_states and name in self._build_app_obj_refs:
             logger.info(
                 f"Received new config deployment for {name} request. Cancelling "
                 "previous request."
             )
-            ray.cancel(self._application_states[name].deploy_obj_ref)
+            ray.cancel(self._build_app_obj_refs[name])
+
+        self._build_app_obj_refs[name] = build_app_obj_ref
         self._application_states[name] = ApplicationState(
             name,
             self.deployment_state_manager,
-            deploy_obj_ref=deploy_obj_ref,
+            self.endpoint_state,
             deployment_time=deployment_time,
         )
 
-    def get_deployment_timestamp(self, name: str) -> float:
+    def deploy_application(
+        self, name: str, deployment_args_list: List[Dict]
+    ) -> List[bool]:
+        """Deploy single application
+
+        Args:
+            name: application name
+            deployment_args: deployment arguments needed for deploying a list of d
+                deployments.
+        Returns:
+            Returns a list indicating whether each deployment is being updated.
+        """
+
+        # Make sure route_prefix is not being used by other application.
+        live_route_prefixes: Dict[str, str] = {
+            self._application_states[app_name].route_prefix: app_name
+            for app_name, app_state in self._application_states.items()
+            if app_state.route_prefix is not None
+            and not app_state.status == ApplicationStatus.DELETING
+            and name != app_name
+        }
+
+        for deploy_param in deployment_args_list:
+            deploy_app_prefix = deploy_param.get("route_prefix")
+            if deploy_app_prefix in live_route_prefixes:
+                raise RayServeException(
+                    f"Prefix {deploy_app_prefix} is being used by application "
+                    f'"{live_route_prefixes[deploy_app_prefix]}".'
+                    f' Failed to deploy application "{name}".'
+                )
+
         if name not in self._application_states:
-            return -1
-        return self._application_states[name].deployment_timestamp
+            self._application_states[name] = ApplicationState(
+                name,
+                self.deployment_state_manager,
+                self.endpoint_state,
+            )
+
+        record_extra_usage_tag(
+            TagKey.SERVE_NUM_APPS, str(len(self._application_states))
+        )
+
+        return self._application_states[name].deploy(deployment_args_list)
```

Comment:
This should not be done in the `deploy_application` method, it should be done in the `update` method (following the same reconciler pattern that `DeploymentStateManager` does. This pattern makes it much easier to reason about the behavior & handle failure scenarios